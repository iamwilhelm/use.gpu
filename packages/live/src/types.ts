// Live function
export type LiveFunction<F extends Function> = (fiber: LiveFiber<F>) => F;

// Mounting key
export type Key = string | number;

// Component with single props object
export type Component<P> = (props: P) => LiveElement<any>;
export type LiveComponent<P> = LiveFunction<Component<P>>;

// State hook callbacks
export type Initial<T> = (() => T) | T;
export type Reducer<T> = T | ((t: T) => T);
export type Setter<T> = (t: Reducer<T>) => void;
export type Resource<T> = () => (void | Task | [T, Task]);

// Deferred actions
export type Task = () => void;
export type Action<F extends Function> = {
  fiber: LiveFiber<F>,
  task: Task,
};
export type Dispatcher = (as: Action<any>[]) => void;
export type OnFiber = (fiber: LiveFiber<any>) => void;
export type FiberSetter<T> = (fiber: LiveFiber<any>, t: T) => void;

// User=defined context
export type LiveContext<T> = { initialValue?: T, displayName?: string };

// Fiber context
export type LiveFiber<F extends Function> = FunctionCall<F> & {
  host?: HostInterface,
  path: Key[],
  depth: number,
	id: number,

  // Instance of F bound to self
  bound?: F,
  
  // State for user hooks
  state: any[],
  pointer: number,

  // State for per-component memoization
  version: number,
  memo: number,

  // Last snapshot
  type: Function,

  // Mounting state
  seen?: Set<Key>,
  mount?: LiveFiber<any> | null,
  mounts?: FiberMap | null,
  order?: Key[],
  next?: LiveFiber<any> | null,

  // User-specified context
  context: FiberContext,

  // Yeeting state
  yeeted?: FiberYeet<any>,
};

export type FiberMap = Map<Key, LiveFiber<any>>;

export type FiberContext = {
  values: ContextValues,
  roots: ContextRoots,
};

export type ContextValues = Map<LiveContext<any>, any>;
export type ContextRoots = Map<LiveContext<any>, LiveFiber<any>>;

export type FiberYeet<T> = {
  id: number,
  emit: FiberSetter<any>,
  value?: T,
  reduced?: T,
  parent?: FiberYeet<T>,
  roots: LiveFiber<any>[],
};

export type GroupedFibers = {
  root: LiveFiber<any>,
  subs: Set<LiveFiber<any>>,
};

// Deferred function calls
export type FunctionCall<F extends Function = Function> = {
  f: LiveFunction<F>,
  args?: any[],
  arg?: any
};

export type DeferredCall<F extends Function = Function> = FunctionCall<F> & {
  key?: Key,
};

export type LiveElement<F = any> = null | DeferredCall<any> | LiveElement<any>[];

// Live host interface
export type HostInterface = {
  // Schedule a task on next flush
  schedule: (fiber: LiveFiber<any>, task: Task) => void,

  // Track a future cleanup on a fiber
  track: (fiber: LiveFiber<any>, task: Task) => void,

  // Dispose of a fiber by running all tracked cleanups
  dispose: (fiber: LiveFiber<any>) => void,

  // Track a long-range dependency for contexts
  depend: (fiber: LiveFiber<any>, root: LiveFiber<any>) => boolean,
  undepend: (fiber: LiveFiber<any>, root: LiveFiber<any>) => void,
  invalidate: (fiber: LiveFiber<any>) => LiveFiber<any>[],

  __stats: {mounts: number, unmounts: number, updates: number, dispatch: number},
  __flush: () => void,
  __ping: (fiber: LiveFiber<any>) => void,
};
