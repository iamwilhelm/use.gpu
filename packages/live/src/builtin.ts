import type {
  Initial, Setter, Reducer, Key, Task,
  LiveFunction, LiveComponent, LiveFiber, LiveCapture, LiveContext, LiveElement, LiveMap, LiveNode,
  FunctionCall, DeferredCall, HostInterface, ArrowFunction,
  ReactElementInterop,
} from './types';

import { compareFibers } from './util';
import { makeFiber, makeImperativeFunction } from './fiber';
import { getCurrentFiberID } from './current';

/** @hidden */
export const MORPH        = () => {};
/** @hidden */
export const DETACH       = () => {};
/** @hidden */
export const FRAGMENT     = () => {};
/** @hidden */
export const MAP_REDUCE   = () => {};
/** @hidden */
export const GATHER       = () => {};
/** @hidden */
export const MULTI_GATHER = () => {};
/** @hidden */
export const FENCE        = () => {};
/** @hidden */
export const YEET         = () => {};
/** @hidden */
export const PROVIDE      = () => {};
/** @hidden */
export const CAPTURE      = () => {};
/** @hidden */
export const DEBUG        = () => {};
/** @hidden */
export const SUSPEND      = () => {};

(MORPH        as any).isLiveBuiltin = true;
(DETACH       as any).isLiveBuiltin = true;
(FRAGMENT     as any).isLiveBuiltin = true;
(MAP_REDUCE   as any).isLiveBuiltin = true;
(GATHER       as any).isLiveBuiltin = true;
(MULTI_GATHER as any).isLiveBuiltin = true;
(FENCE        as any).isLiveBuiltin = true;
(YEET         as any).isLiveBuiltin = true;
(PROVIDE      as any).isLiveBuiltin = true;
(CAPTURE      as any).isLiveBuiltin = true;
(DEBUG        as any).isLiveBuiltin = true;
(SUSPEND      as any).isLiveBuiltin = true;

(FRAGMENT     as any).isLiveInline = true;

// Inline render ops
type UseArgs<F> = F extends ArrowFunction ? Parameters<F> : any[];

interface Use<F extends ArrowFunction> {
  (f: LiveFunction<F>): DeferredCall<F>;
  (f: LiveFunction<F>, ...args: UseArgs<F>): DeferredCall<F>;
};

/** Use a call to a Live function, reconciled by numeric index. */
export const use: Use<any> = <F extends ArrowFunction>(
  f: LiveFunction<F>,
  ...args: UseArgs<F>
): DeferredCall<F> => ({f, args, key: undefined, by: getCurrentFiberID()} as any);

/** Use a keyed call to a Live function, reconciled by key. */
export const keyed = <F extends ArrowFunction>(
  f: LiveFunction<F>,
  key?: Key,
  ...args: UseArgs<F>
): DeferredCall<F> => ({f, args, key, by: getCurrentFiberID()} as any);

/** Use a call to a Live Component with only a children prop. */
export const wrap = <F extends ArrowFunction>(
  f: LiveFunction<F>,
  children: any,
): DeferredCall<F> => ({f, args: [{children}], key: undefined, by: getCurrentFiberID()} as any);

/** Add arguments to an existing call or calls. */
export const extend = (
  calls: LiveNode<any>,
  props: Record<string, any>,
): LiveElement<any> => {
  if (typeof calls === 'string') return null;
  if (typeof calls === 'function') return null;
  if (!calls) return calls;

  if (Array.isArray(calls)) return calls.map(call => extend(call, props)) as any;
  if ('props' in calls) {
    return extend(keyed(calls.type, calls.key, calls.props), props);
  }
  else if (calls.args?.length) {
    const [existing, ...rest] = calls.args;
    return ({...calls, args: [{...existing, ...props}, ...rest] });
  }
  if (!calls) return calls;
  return ({...calls, args: [props] } as any);
}

/** Morph a call to a Live function.

Morphing will change a component's type, without forcibly unmounting its children.

Children are still reconciled as usual, and only stay if re-rendered by the new parent.

The parent that is being morphed still loses all its own state.
*/
export const morph = (
  calls: LiveNode<any>,
  key?: Key,
): DeferredCall<() => void> => ({f: MORPH, args: calls as any, key, by: getCurrentFiberID()} as any);

/** Detach the rendering of a Live subtree.

The callback is invoked with a render function, which it can call repeatedly to render the detached fiber. */
export const detach = <F extends ArrowFunction>(
  call: DeferredCall<F>,
  callback: (render: () => void, fiber: LiveFiber<F>) => void,
  key?: Key,
): DeferredCall<() => void> => ({f: DETACH, args: [call, callback], key, by: getCurrentFiberID()} as any);

/** Holds an array of Live calls to reconcile. */
export const fragment = (
  calls: LiveNode<any>,
  key?: Key,
): DeferredCall<() => void> => {
  if (Array.isArray(calls)) return calls as any;
  return [calls] as any;
}

/** Wrap a fragment in a debug node to mark it. */
export const debug = (
  calls: LiveNode<any>,
  key?: Key,
): DeferredCall<() => void> => {
  if (Array.isArray(calls)) return ({f: DEBUG, args: calls, key, by: getCurrentFiberID()} as any);
  return ({f: DEBUG, args: [calls], key} as any);
}

/** Reduce values from a subtree, using the given `map(...)` and `reduce(...)`. */
export const mapReduce = <R, T>(
  calls?: LiveNode<any>,
  map?: (t: T) => R,
  reduce?: (a: R, b: R) => R,
  then?: LiveFunction<(r: R) => LiveElement<any>>,
  key?: Key,
): DeferredCall<() => void> => ({f: MAP_REDUCE, args: [calls, map, reduce, then], key, by: getCurrentFiberID()} as any);

/** Gather items from a subtree, into a flat array. */
export const gather = <T>(
  calls?: LiveNode<any>,
  then?: LiveFunction<(r: T[]) => LiveElement<any>>,
  key?: Key,
): DeferredCall<() => void> => ({f: GATHER, args: [calls, then], key, by: getCurrentFiberID()} as any);

/** Multi-gather items from a subtree, by object key. */
export const multiGather = <T>(
  calls?: LiveNode<any>,
  then?: LiveFunction<(r: Record<string, T[]>) => LiveElement<any>>,
  key?: Key,
): DeferredCall<() => void> => ({f: MULTI_GATHER, args: [calls, then], key, by: getCurrentFiberID()} as any);

/** Fence gathered items from a subtree. */
export const fence = <T>(
  calls?: LiveNode<any>,
  then?: LiveFunction<(r: T[]) => LiveElement<any>>,
  key?: Key,
): DeferredCall<() => void> => ({f: FENCE, args: [calls, then], key, by: getCurrentFiberID()} as any);

/** Yeet value(s) upstream. */
export const yeet = <T>(
  value?: T,
  key?: Key,
): DeferredCall<() => void> => ({f: YEET, arg: value, key, by: getCurrentFiberID()} as any);

/** Provide a value for a Live context. */
export const provide = <T, C>(
  context: LiveContext<C>,
  value: T,
  calls?: LiveNode<any>,
  key?: Key,
): DeferredCall<() => void> => ({f: PROVIDE, args: [context, value, calls], key, by: getCurrentFiberID()} as any);

/** Capture values from a Live context. */
export const capture = <T, C>(
  context: LiveCapture<C>,
  calls?: LiveNode<any>,
  then?: LiveFunction<(r: T) => void>,
  key?: Key,
): DeferredCall<() => void> => ({f: CAPTURE, args: [context, calls, then], key, by: getCurrentFiberID()} as any);

/** Component has side-effects, and will re-render even if props object is identical. */
export const imperative = makeImperativeFunction;

/** Yeet a suspend symbol. */
export const suspend = () => yeet(SUSPEND);

/** LOL. Look, _you_ go try to make JSX.Element polymorphic. */
export const into = (children: any): any => children;

export interface MakeContext<T> {
  <T>(initialValue: T, displayName?: string): LiveContext<T>;
  <T>(initialValue: undefined, displayName?: string): LiveContext<T>;
  <T>(initialValue: null, displayName?: string): LiveContext<T | null>;
};

/** Make Live context for holding shared value for child nodes (defaulted, required or optional). */
export const makeContext: MakeContext<unknown> = <T>(initialValue?: T | null, displayName?: string) => ({
  initialValue,
  displayName,
  context: true,
});

/** Make Live capture for holding shared value for child nodes (defaulted, required or optional). */
export const makeCapture = <T>(displayName?: string): LiveCapture<T> => ({
  displayName,
  capture: true,
});

/** Flatten a captured value map into an ordered array. */
export const captureFibers = <T>(map: Map<LiveFiber<any>, T>): [LiveFiber<any>, T][] => {
  const entries = Array.from(map.entries());
  entries.sort((a, b) => compareFibers(a[0], b[0]));
  return entries;
}

/** Flatten a captured value map into an ordered array. */
export const captureValues = <T>(
  map: LiveMap<T>,
): T[] => {
  const keys = Array.from(map.keys());
  keys.sort((a, b) => compareFibers(a, b));
  return keys.map(k => map.get(k)!);
}

/** Get last node of captured value map. */
export const captureLastFiber = <T>(
  map: LiveMap<T>,
): [LiveFiber<any>, T] | null => {
  const keys = Array.from(map.keys());
  keys.sort((a, b) => compareFibers(a, b));
  
  const key = keys[keys.length - 1];
  return key ? [key, map.get(key)!] : null;
}

/** Get last value yielded from captured value map. */
export const captureTail = <T>(
  map: LiveMap<T>
): T | null => captureLastFiber(map)?.[1] ?? null;
