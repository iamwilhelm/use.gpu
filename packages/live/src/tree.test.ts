import { LiveFiber, Task } from './types';
import { use, detach, provide, provideMemo, PROVIDE, makeContext } from './builtin';
import { renderFiber } from './fiber';
import { memoArgs, useState, useContext } from './hooks';
import { renderSync } from './tree';

it("mounts", () => {

  const Root = () => () => use(Node)();
  const Node = () => () => {};

  const result = renderSync(use(Root)());

  expect(result.host).toBeTruthy();
  if (!result.host) return;

  expect(result.f).toBe(Root);
  expect(result.mount).toBeTruthy();
  if (!result.mount) return;

  const node = result.mount;
  expect(node && node.f).toBe(Node);
});

it("mounts multiple", () => {

  const Root = () => () => [
    use(Node, '1')(),
    use(Node, '2')(),
  ];

  const Node = () => () => {};

  const result = renderSync(use(Root)());
  expect(result.f).toBe(Root);
  expect(result.mounts).toBeTruthy();
  if (!result.mounts) return;

  const node1 = result.mounts.get('1');
  const node2 = result.mounts.get('2');
  expect(node1 && node1.f).toBe(Node);
  expect(node2 && node2.f).toBe(Node);
});

it("mounts a subfiber", () => {

  let captureSubFiber: LiveFiber<any> | null = null;

  const Root = (fiber: LiveFiber<any>) => () =>
    detach(use(Sub)(), (render: () => void, mount: LiveFiber<any>) => {
      captureSubFiber = mount;
      render();
    });

  const Sub = () => () => use(Node)();
  const Node = () => () => {};

  const result = renderSync(use(Root)());
  expect(result.f).toBe(Root);

  expect(result.mount).toBeTruthy();
  expect(result.mount!.next).toBeTruthy();

  expect(captureSubFiber).toBeTruthy();
  if (captureSubFiber != null) {
    const {mount} = captureSubFiber;
    // @ts-ignore
    expect(mount && mount.f).toBe(Node);
  }

});

it("reacts on the root (setter form)", () => {

  const rendered = {
    root: 0,
    node: 0,
  };
  let trigger = null as Task | null;
  const setTrigger = (f: Task) => trigger = f;

  const Root = (fiber: LiveFiber<any>) => () => {
    rendered.root++;

    const [, setValue] = useState(0);
    setTrigger(() => setValue(1));

    return use(Node)(Math.random());
  };

  const Node = () => (x?: number) => {
    rendered.node++;
  };

  const result = renderSync(use(Root)());
  expect(result.host).toBeTruthy();
  expect(result.mount).toBeTruthy();
  if (!result.host) return;
  if (!result.mount) return;

  const {host: {__flush: flush}} = result;

  expect(result.f).toBe(Root);
  const node1 = result.mount;
  expect(node1 && node1.f).toBe(Node);

  expect(rendered.root).toBe(1);
  expect(rendered.node).toBe(1);

  if (trigger) trigger();
  if (flush) flush();

  expect(result.f).toBe(Root);
  const node2 = result.mount;
  expect(node2 && node2.f).toBe(Node);

  expect(rendered.root).toBe(2);
  expect(rendered.node).toBe(2);

});

it("reacts on the root (reducer form)", () => {

  const rendered = {
    root: 0,
    node: 0,
  };
  let trigger = null as Task | null;
  const setTrigger = (f: Task) => trigger = f;

  const Root = (fiber: LiveFiber<any>) => () => {
    rendered.root++;

    const [, setValue] = useState(0);
    setTrigger(() => setValue((s: number) => s + 1));

    return use(Node)(Math.random());
  };

  const Node = () => (x?: number) => {
    rendered.node++;
  };

  const result = renderSync(use(Root)());
  expect(result.host).toBeTruthy();
  expect(result.mount).toBeTruthy();
  if (!result.host) return;
  if (!result.mount) return;

  const {host: {__flush: flush}} = result;

  expect(result.f).toBe(Root);
  const node1 = result.mount;
  expect(node1 && node1.f).toBe(Node);

  expect(rendered.root).toBe(1);
  expect(rendered.node).toBe(1);

  if (trigger) trigger();
  if (flush) flush();

  expect(result.f).toBe(Root);
  const node2 = result.mount;
  expect(node2 && node2.f).toBe(Node);

  expect(rendered.root).toBe(2);
  expect(rendered.node).toBe(2);

});

it("reacts and remounts on the root", () => {

  const rendered = {
    root: 0,
    node: 0,
  };
  let trigger = null as Task | null;
  const setTrigger = (f: Task) => trigger = f;

  const Root = (fiber: LiveFiber<any>) => () => {
    const [, setValue] = useState(0);
    setTrigger(() => setValue(1));

    rendered.root++;
    return [
      use(Node, '1')(Math.random()),
      use(Node, '2')(Math.random()),
      use(Node, '3' + rendered.root)(Math.random()),
    ];
  };

  const Node = () => (x?: number) => {
    rendered.node++;
  };

  const result = renderSync(use(Root)());
  expect(result.host).toBeTruthy();
  if (!result.host) return;

  const {host: {__flush: flush, __stats: stats}} = result;

  expect(result.f).toBe(Root);
  expect(result.mounts).toBeTruthy();
  if (result.mounts) {
    const node1 = result.mounts.get('1');
    const node2 = result.mounts.get('2');
    const node3 = result.mounts.get('31');
    const node4 = result.mounts.get('32');
    expect(node1 && node1.f).toBe(Node);
    expect(node2 && node2.f).toBe(Node);
    expect(node3 && node3.f).toBe(Node);
    expect(node4).toBe(undefined);
  }

  expect(rendered.root).toBe(1);
  expect(rendered.node).toBe(3);

  expect(stats.mounts).toBe(4);
  expect(stats.unmounts).toBe(0);
  expect(stats.updates).toBe(0);
  expect(stats.dispatch).toBe(1);

  if (trigger) trigger();
  if (flush) flush();

  expect(result.f).toBe(Root);
  expect(result.mounts).toBeTruthy();
  if (result.mounts) {
    const node1 = result.mounts.get('1');
    const node2 = result.mounts.get('2');
    const node3 = result.mounts.get('31');
    const node4 = result.mounts.get('32');
    expect(node1 && node1.f).toBe(Node);
    expect(node2 && node2.f).toBe(Node);
    expect(node3).toBe(undefined);
    expect(node4 && node4.f).toBe(Node);
  }

  expect(rendered.root).toBe(2);
  expect(rendered.node).toBe(6);

  expect(stats.mounts).toBe(5);
  expect(stats.unmounts).toBe(1);
  expect(stats.updates).toBe(2);
  expect(stats.dispatch).toBe(2);

});

it("reacts and remounts a sub tree", () => {

  const rendered = {
    root: 0,
    subroot: 0,
    node: 0,
  };
  let trigger = null as Task | null;
  const setTrigger = (f: Task) => trigger = f;

  const Root = () => () => {
    rendered.root++;
    return [
      use(SubRoot, 'subroot')(),
    ];
  };

  const SubRoot = (fiber: LiveFiber<any>) => () => {
    const [, setValue] = useState(0);
    setTrigger(() => setValue(1));

    rendered.subroot++;
    return [
      use(Node, '1')(Math.random()),
      use(Node, '2')(Math.random()),
      use(Node, '3' + rendered.subroot)(Math.random()),
    ];
  };

  const Node = () => (x?: number) => {
    rendered.node++;
    return;
  };

  const result = renderSync(use(Root)());
  expect(result.host).toBeTruthy();
  expect(result.mounts).toBeTruthy();
  if (!result.host) return;
  if (!result.mounts) return;

  const {host: {__flush: flush, __stats: stats}} = result;

  expect(result.f).toBe(Root);
  expect(result.mounts).toBeTruthy();

  const sub1 = result.mounts.get('subroot');
  const node1 = sub1 && sub1.mounts && sub1.mounts.get('31');
  expect(sub1 && sub1.f).toBe(SubRoot);
  expect(node1 && node1.f).toBe(Node);

  expect(rendered.root).toBe(1);
  expect(rendered.subroot).toBe(1);
  expect(rendered.node).toBe(3);

  expect(stats.mounts).toBe(5);
  expect(stats.unmounts).toBe(0);
  expect(stats.updates).toBe(0);
  expect(stats.dispatch).toBe(1);

  if (trigger) trigger();
  if (flush) flush();

  expect(result.f).toBe(Root);
  expect(result.mounts).toBeTruthy();

  const sub2 = result.mounts.get('subroot');
  const node2 = sub2 && sub2.mounts && sub2.mounts.get('32');
  expect(sub2 && sub2.f).toBe(SubRoot);
  expect(node2 && node2.f).toBe(Node);

  expect(rendered.root).toBe(1);
  expect(rendered.subroot).toBe(2);
  expect(rendered.node).toBe(6);

  expect(stats.mounts).toBe(6);
  expect(stats.unmounts).toBe(1);
  expect(stats.updates).toBe(2);
  expect(stats.dispatch).toBe(2);

});

it("coalesces updates", () => {

  const rendered = {
    root: 0,
    node: 0,
  };
  let trigger1 = null as Task | null;
  let trigger2 = null as Task | null;
  const setTrigger1 = (f: Task) => trigger1 = f;
  const setTrigger2 = (f: Task) => trigger2 = f;

  const Root = (fiber: LiveFiber<any>) => () => {
    rendered.root++;

    const [, setValue] = useState(0);
    setTrigger1(() => setValue(1));

    return use(Node)(Math.random());
  };

  const Node = (fiber: LiveFiber<any>) => (x?: number) => {
    rendered.node++;

    const [, setValue] = useState(0);
    setTrigger2(() => setValue(1));
  };

  const result = renderSync(use(Root)());
  expect(result.host).toBeTruthy();
  expect(result.mount).toBeTruthy();
  if (!result.host) return;
  if (!result.mount) return;

  const {host: {__flush: flush}} = result;

  expect(result.f).toBe(Root);
  const node1 = result.mount;
  expect(node1 && node1.f).toBe(Node);

  expect(rendered.root).toBe(1);
  expect(rendered.node).toBe(1);

  if (trigger1) trigger1();
  if (trigger2) trigger2();
  if (flush) flush();

  expect(result.f).toBe(Root);
  const node2 = result.mount;
  expect(node2 && node2.f).toBe(Node);

  expect(rendered.root).toBe(2);
  expect(rendered.node).toBe(2);

});

it("updates with memo in the way", () => {

  const rendered = {
    root: 0,
    memo: 0,
    node: 0,
  };
  let trigger1 = null as Task | null;
  let trigger2 = null as Task | null;
  const setTrigger1 = (f: Task) => trigger1 = f;
  const setTrigger2 = (f: Task) => trigger2 = f;

  const Root = (fiber: LiveFiber<any>) => () => {
    rendered.root++;

    const [, setValue] = useState(0);
    setTrigger1(() => setValue(1));

    return use(Memo)();
  };

  const Memo = memoArgs((fiber: LiveFiber<any>) => () => {
    rendered.memo++;

    return use(Node)(Math.random());
  });

  const Node = (fiber: LiveFiber<any>) => (x?: number) => {
    rendered.node++;

    const [, setValue] = useState(0);
    setTrigger2(() => setValue(1));
  };

  const result = renderSync(use(Root)());
  expect(result.host).toBeTruthy();
  expect(result.mount).toBeTruthy();
  expect(result.mount!.mount).toBeTruthy();
  if (!result.host) return;
  if (!result.mount) return;
  if (!result.mount.mount) return;

  const {host: {__flush: flush}} = result;

  expect(result.f).toBe(Root);
  const memo1 = result.mount;
  expect(memo1 && memo1.f).toBe(Memo);
  const node1 = result.mount.mount;
  expect(node1 && node1.f).toBe(Node);

  expect(rendered.root).toBe(1);
  expect(rendered.memo).toBe(1);
  expect(rendered.node).toBe(1);

  if (trigger1) trigger1();
  if (trigger2) trigger2();
  if (flush) flush();

  expect(result.f).toBe(Root);
  const memo2 = result.mount;
  expect(memo2 && memo2.f).toBe(Memo);
  const node2 = result.mount.mount;
  expect(node2 && node2.f).toBe(Node);

  expect(rendered.root).toBe(2);
  expect(rendered.memo).toBe(1);
  expect(rendered.node).toBe(2);

});

it("updates context with memo in the way", () => {

  const context = makeContext();

  const rendered = {
    root: 0,
    memo: 0,
    node: 0,
    value: -1,
  };
  let trigger = null as Task | null;
  const setTrigger = (f: Task) => trigger = f;

  const Root = (fiber: LiveFiber<any>) => () => {
    rendered.root++;

    const [value, setValue] = useState(0);
    setTrigger(() => setValue(1));

    return provide(context, value, use(Memo)());
  };

  const Memo = memoArgs((fiber: LiveFiber<any>) => () => {
    rendered.memo++;

    return use(Node)();
  });

  const Node = (fiber: LiveFiber<any>) => () => {
    rendered.node++;
    const value = useContext(context);
    rendered.value = value;
  };

  const result = renderSync(use(Root)());
  expect(result.host).toBeTruthy();
  expect(result.mount).toBeTruthy();
  expect(result.mount!.mount).toBeTruthy();
  if (!result.host) return;
  if (!result.mount) return;
  if (!result.mount.mount) return;

  const {host: {__flush: flush}} = result;

  expect(result.f).toBe(Root);
  const provide1 = result.mount;
  expect(provide1 && provide1.f).toBe(PROVIDE);
  const memo1 = result.mount.mount;
  expect(memo1 && memo1.f).toBe(Memo);
  const node1 = result.mount.mount.mount;
  expect(node1 && node1.f).toBe(Node);

  expect(rendered.root).toBe(1);
  expect(rendered.memo).toBe(1);
  expect(rendered.node).toBe(1);
  expect(rendered.value).toBe(0);

  if (trigger) trigger();
  if (flush) flush();

  expect(result.f).toBe(Root);
  const provide2 = result.mount;
  expect(provide2 && provide2.f).toBe(PROVIDE);
  const memo2 = result.mount.mount;
  expect(memo2 && memo2.f).toBe(Memo);
  const node2 = result.mount.mount.mount;
  expect(node2 && node2.f).toBe(Node);

  expect(rendered.root).toBe(2);
  expect(rendered.memo).toBe(1);
  expect(rendered.node).toBe(2);
  expect(rendered.value).toBe(1);

});

it("does update context if value is the same", () => {

  const context = makeContext();

  const rendered = {
    root: 0,
    memo: 0,
    node: 0,
    value: -1,
  };
  let trigger = null as Task | null;
  const setTrigger = (f: Task) => trigger = f;

  const Root = (fiber: LiveFiber<any>) => () => {
    rendered.root++;

    const [value, setValue] = useState(0);
    setTrigger(() => setValue(1));

    return provide(context, 0, use(Memo)());
  };

  const Memo = memoArgs((fiber: LiveFiber<any>) => () => {
    rendered.memo++;

    return use(Node)();
  });

  const Node = (fiber: LiveFiber<any>) => () => {
    rendered.node++;
    const value = useContext(context);
    rendered.value = value;
  };

  const result = renderSync(use(Root)());
  expect(result.host).toBeTruthy();
  expect(result.mount).toBeTruthy();
  expect(result.mount!.mount).toBeTruthy();
  if (!result.host) return;
  if (!result.mount) return;
  if (!result.mount.mount) return;

  const {host: {__flush: flush}} = result;

  expect(result.f).toBe(Root);
  const provide1 = result.mount;
  expect(provide1 && provide1.f).toBe(PROVIDE);
  const memo1 = result.mount.mount;
  expect(memo1 && memo1.f).toBe(Memo);
  const node1 = result.mount.mount.mount;
  expect(node1 && node1.f).toBe(Node);

  expect(rendered.root).toBe(1);
  expect(rendered.memo).toBe(1);
  expect(rendered.node).toBe(1);
  expect(rendered.value).toBe(0);

  if (trigger) trigger();
  if (flush) flush();

  expect(result.f).toBe(Root);
  const provide2 = result.mount;
  expect(provide2 && provide2.f).toBe(PROVIDE);
  const memo2 = result.mount.mount;
  expect(memo2 && memo2.f).toBe(Memo);
  const node2 = result.mount.mount.mount;
  expect(node2 && node2.f).toBe(Node);

  expect(rendered.root).toBe(2);
  expect(rendered.memo).toBe(1);
  expect(rendered.node).toBe(2);
  expect(rendered.value).toBe(0);

});

it("does not update memoized context if value is the same", () => {

  const context = makeContext();

  const rendered = {
    root: 0,
    memo: 0,
    node: 0,
    value: -1,
  };
  let trigger = null as Task | null;
  const setTrigger = (f: Task) => trigger = f;

  const Root = (fiber: LiveFiber<any>) => () => {
    rendered.root++;

    const [value, setValue] = useState(0);
    setTrigger(() => setValue(1));

    return provideMemo(context, 0, memoChild);
  };

  const Memo = memoArgs((fiber: LiveFiber<any>) => () => {
    rendered.memo++;

    return use(Node)();
  });

  const Node = (fiber: LiveFiber<any>) => () => {
    rendered.node++;
    const value = useContext(context);
    rendered.value = value;
  };

  const memoChild = use(Memo)();

  const result = renderSync(use(Root)());
  expect(result.host).toBeTruthy();
  expect(result.mount).toBeTruthy();
  expect(result.mount!.mount).toBeTruthy();
  if (!result.host) return;
  if (!result.mount) return;
  if (!result.mount.mount) return;

  const {host: {__flush: flush}} = result;

  expect(result.f).toBe(Root);
  const provide1 = result.mount;
  expect(provide1 && provide1.f).toBe(PROVIDE);
  const memo1 = result.mount.mount;
  expect(memo1 && memo1.f).toBe(Memo);
  const node1 = result.mount.mount.mount;
  expect(node1 && node1.f).toBe(Node);

  expect(rendered.root).toBe(1);
  expect(rendered.memo).toBe(1);
  expect(rendered.node).toBe(1);
  expect(rendered.value).toBe(0);

  if (trigger) trigger();
  if (flush) flush();

  expect(result.f).toBe(Root);
  const provide2 = result.mount;
  expect(provide2 && provide2.f).toBe(PROVIDE);
  const memo2 = result.mount.mount;
  expect(memo2 && memo2.f).toBe(Memo);
  const node2 = result.mount.mount.mount;
  expect(node2 && node2.f).toBe(Node);

  expect(rendered.root).toBe(2);
  expect(rendered.memo).toBe(1);
  expect(rendered.node).toBe(1);
  expect(rendered.value).toBe(0);

});
