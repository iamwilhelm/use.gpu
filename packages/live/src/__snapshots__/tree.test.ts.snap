// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`captures values 1`] = `
"<Root >
  <Capture(undefined) displayName={undefined} capture={true} [<Node >,<Node >,…] (…) (0, _builtin.use)(Node, (0, _builtin.ca…>
    <Node >
    <Node >
    <Node >
    <Resume(undefined) (…) (0, _builtin.use)(Node, (0, _builtin.ca…>
      <Node [1,2,…]>"
`;

exports[`gathers yeeted values 1`] = `
"<Root >
  <Yeet 1>
  <Value >
  <Yeet 3>
  <Resume(Root) (…) (0, _builtin.use)(Node, values)>
    <Node [1,2,…]>"
`;

exports[`render reordering 1`] = `
"<Root >
  <Provide(undefined) initialValue={-1} displayName={undefined} context={true} 0 <Memo() >>
    <Memo() >
      <Node id={0}>
      <Node id={1}>
      <Node id={2}>"
`;

exports[`render reordering 2`] = `
"<Root >
  <Provide(undefined) initialValue={-1} displayName={undefined} context={true} 1 <Memo() >>
    <Memo() >
      <Node id={1}>
      <Node id={2}>
      <Node id={0}>"
`;

exports[`renders quote/unquote pairs 1`] = `
"<Root >
  <First <Quote <Second …>>>
    <Quote <Second <Unquote …>>>
      <First <Quote <Unquote …>>>
        <First <First <Quote …>>>
          <First <Quote <Second >>>
        <Reconcile(First) >
      <Reconcile(Quote) >
    <Reconcile(First) >
  <Root(…) >
    <Second <Second <Unquote …>>>
      <Second <Unquote <Quote …>>>
        <Second <Unquote <First …>>>
          <Unquote <First <First …>>>
            <Second >
            <Reconcile(Unquote) >
          <Reconcile(Second) >
        <Reconcile(Second) >
    <Reconcile(…) >"
`;

exports[`renders quoted tree 1`] = `
"<Root >
  <Node >
  <Quote <Tree >>
  <Node >
  <Root(…) >
    <Tree >
      <Node <Node <Node >>>
        <Node <Node >>
          <Node >
    <Reconcile(…) >"
`;

exports[`renders quoted/unquoted trees 1`] = `
"<Root >
  <Reconcile <Quote <Second …>>>
    <Quote <Second <Second …>>>
      <First <Quote <Second …>>>
        <First <First <Quote …>>>
          <First <Quote <Second >>>
        <Reconcile(First) >
      <Reconcile(Quote) >
    <Root(Reconcile) >
      <Second <Second <Unquote …>>>
        <Second <Unquote <First …>>>
          <Second <Unquote <First …>>>
            <Second >
            <Reconcile(Second) >
          <Reconcile(Second) >
      <Reconcile(…) >"
`;

exports[`yeets from capture 1`] = `
"<Root >
  <Capture(undefined) displayName={undefined} capture={true} [f={YEET(…) {}} arg={1} key={undefined} by={51},f={…} arg={…} key={…} by={…},…,…] (…) (0, _builtin.yeet)((0, _builtin.capture…>
    <Yeet 1>
    <Yeet 2>
    <Node >
    <Node >
    <Resume(undefined) (…) (0, _builtin.yeet)((0, _builtin.capture…>
  <Resume(Root) (…) (0, _builtin.use)(Node, values)>
    <Node [1,2,…,…]>"
`;
